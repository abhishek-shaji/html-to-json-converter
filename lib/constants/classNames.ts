export const CLASS_NAME_CHARACTER = 'character';
export const CLASS_NAME_DIALOG = 'dialog';
