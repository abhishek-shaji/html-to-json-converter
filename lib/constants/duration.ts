export const TIME_BETWEEN_CHAR_MS = 70;
