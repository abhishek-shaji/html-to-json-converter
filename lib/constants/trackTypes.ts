export const TRACK_TYPE_DIALOGUE = 'DI';
export const TRACK_TYPE_ANIMATION = 'AN';
export const TRACK_TYPE_EMOTION = 'EM';
