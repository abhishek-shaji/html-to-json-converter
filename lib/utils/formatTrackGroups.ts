import { OutputType, TrackType } from '../types/output';

const formatTracks = (trackMap): TrackType[] =>
  Object.keys(trackMap).reduce((accum: TrackType[], type): TrackType[] => {
    return [
      ...accum,
      {
        type,
        actions: trackMap[type],
      },
    ];
  }, []);

export const formatTrackGroups = (trackGroupData): OutputType =>
  Object.keys(trackGroupData).reduce(
    (accum: OutputType, target): OutputType => {
      return {
        trackGroups: [
          ...accum.trackGroups,
          {
            target,
            tracks: formatTracks(trackGroupData[target]),
          },
        ],
      };
    },
    {
      trackGroups: [],
    }
  );
