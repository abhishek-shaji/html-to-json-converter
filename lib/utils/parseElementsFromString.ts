import { JSDOM } from 'jsdom';
import { EmptyFileException } from '../exceptions/EmptyFileException';

export const parseElementsFromString = (html: string): NodeListOf<Element> => {
  if (!html.trim()) {
    throw new EmptyFileException();
  }

  const dom = new JSDOM(html);
  const { document } = dom.window;

  return document.querySelectorAll('div.character, div.dialog');
};
