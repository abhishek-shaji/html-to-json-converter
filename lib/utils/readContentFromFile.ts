import { readFileSync } from 'fs';
import { FileReadFailureException } from '../exceptions/FileReadFailureException';

export const readContentFromFile = (path: string): string => {
  try {
    return readFileSync(path, 'utf8');
  } catch (err) {
    console.error(err);

    throw new FileReadFailureException();
  }
};
