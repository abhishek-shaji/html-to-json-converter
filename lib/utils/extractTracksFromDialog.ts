import { TIME_BETWEEN_CHAR_MS } from '../constants/duration';
import { TRACK_TYPE_DIALOGUE } from '../constants/trackTypes';

export const extractTracksFromDialog = (
  dialog: Element,
  initialState: any,
  timestampStartAt: number
) => {
  const trackMap = { ...initialState };
  const children = [...dialog.children];

  for (const childElement of children) {
    const actionType = childElement.getAttribute('data-action-type');
    const target = childElement.getAttribute('data-action-target');
    const actionId = childElement.getAttribute('data-action-id');
    const timestamp =
      timestampStartAt + dialog.innerHTML.indexOf(childElement.outerHTML) * TIME_BETWEEN_CHAR_MS;

    if (!trackMap[actionType]) {
      trackMap[actionType] = [];
    }

    trackMap[actionType].push({ actionId, actionType, timestamp, target });

    childElement.remove();
  }

  const dialogueDuration = dialog.textContent.length * TIME_BETWEEN_CHAR_MS;

  if (!trackMap[TRACK_TYPE_DIALOGUE]) {
    trackMap[TRACK_TYPE_DIALOGUE] = [];
  }

  trackMap[TRACK_TYPE_DIALOGUE].push({
    dialog: dialog.textContent,
    duration: dialogueDuration,
    timestamp: timestampStartAt,
  });

  return { trackMap, dialogueDuration };
};
