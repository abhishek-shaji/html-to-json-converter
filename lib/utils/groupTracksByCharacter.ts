import { CLASS_NAME_DIALOG } from '../constants/classNames';
import { extractTracksFromDialog } from './extractTracksFromDialog';

export const groupTracksByCharacter = (elements: NodeListOf<Element>) => {
  const targets = {};
  let offset = 0;

  for (const [elementIndex, element] of elements.entries()) {
    const isDialog = element.classList.contains(CLASS_NAME_DIALOG);

    if (isDialog) {
      const character = elements[elementIndex - 1].innerHTML;

      if (!targets[character]) {
        targets[character] = {};
      }

      const { trackMap, dialogueDuration } = extractTracksFromDialog(
        element,
        targets[character],
        offset
      );

      offset = offset + dialogueDuration;

      targets[character] = trackMap;
    }
  }

  return targets;
};
