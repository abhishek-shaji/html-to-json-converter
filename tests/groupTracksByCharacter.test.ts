import { parseElementsFromString } from '../lib/utils/parseElementsFromString';
import { groupTracksByCharacter } from '../lib/utils/groupTracksByCharacter';
import { TRACK_TYPE_DIALOGUE } from '../lib/constants/trackTypes';

const CHARACTER_NAME_LUCY = 'lucy';
const LUCY_DIALOG = 'Lorem ipsum dolor sit ame';
const CHARACTER_NAME_TIM = 'tim';
const TIM_DIALOG = 'nisi ut aliquip ex ea';

const mockContent = `
  <div class='character'>${CHARACTER_NAME_LUCY}</div>
  <div class='dialog'>${LUCY_DIALOG}</div>
  <div class='character'>${CHARACTER_NAME_TIM}</div>
  <div class='dialog'>${TIM_DIALOG}</div>
`;

describe('groupTracksByCharacter', () => {
  it('should properly group dialog and action data by character name', () => {
    const elements = parseElementsFromString(mockContent);

    const response = groupTracksByCharacter(elements);

    expect(response).toMatchSnapshot();
    expect(response[CHARACTER_NAME_LUCY][TRACK_TYPE_DIALOGUE].length).toEqual(1);
    expect(response[CHARACTER_NAME_LUCY][TRACK_TYPE_DIALOGUE][0].dialog).toEqual(LUCY_DIALOG);
    expect(response[CHARACTER_NAME_LUCY][TRACK_TYPE_DIALOGUE][0].timestamp).toEqual(0);
    expect(response[CHARACTER_NAME_LUCY][TRACK_TYPE_DIALOGUE][0].duration).toEqual(1750);

    expect(response[CHARACTER_NAME_TIM][TRACK_TYPE_DIALOGUE].length).toEqual(1);
    expect(response[CHARACTER_NAME_TIM][TRACK_TYPE_DIALOGUE][0].dialog).toEqual(TIM_DIALOG);
    expect(response[CHARACTER_NAME_TIM][TRACK_TYPE_DIALOGUE][0].timestamp).toEqual(1750);
    expect(response[CHARACTER_NAME_TIM][TRACK_TYPE_DIALOGUE][0].duration).toEqual(1470);
  });
});
