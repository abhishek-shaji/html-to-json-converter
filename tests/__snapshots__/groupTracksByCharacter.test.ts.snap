// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groupTracksByCharacter should properly group dialog and action data by character name 1`] = `
{
  "lucy": {
    "DI": [
      {
        "dialog": "Lorem ipsum dolor sit ame",
        "duration": 1750,
        "timestamp": 0,
      },
    ],
  },
  "tim": {
    "DI": [
      {
        "dialog": "nisi ut aliquip ex ea",
        "duration": 1470,
        "timestamp": 1750,
      },
    ],
  },
}
`;
